app/User.php

```diff
@@ -337,4 +342,121 @@ public static function getBlogModeratorEmails(): array
         return $emails;
     }
 
+    public function lastLogin()
+    {
+        return $this->hasOne(UserLogin::class)->latest();
+    }
+
+    public function isIdle(int $inactiveMonths = 1)
+    {
+
+        $now = new DateTime('NOW');
+        $lastLogin = DateTime::createFromFormat('Y-m-d H:s:i', $this->lastLogin->created_at);
+
+        return (($now->diff($lastLogin)->m) >= $inactiveMonths);
+
+    }
+
+    /**
+     * Check inactive users and sends deactivation notice email to inactive users
+     *
+     * @param int $inactiveMonths
+     * @param string $userRole
+     */
+    public static function sendDeactivationNotice(int $inactiveMonths = 1, string $userRole = self::ROLE_CONTRIBUTOR)
+    {
+
+        $users = self::where('is_active', 1)
+            ->where('is_flagged_for_deactivation', 0)
+            ->where('role', $userRole)
+            ->get();
+
+        foreach ($users as $user) {
+
+            if ($user->isIdle($inactiveMonths)) {
+
+                $user->is_flagged_for_deactivation = 1;
+                $user->deactivation_notice_sent_at = date('Y-m-d H:s:i');
+                $user->save();
+
+                self::sendDeactivationEmail($user);
+
+            }
+
+        }
+
+    }
+
+    /**
+     * Sends deactivation notice email to inactive users
+     *
+     * @param User $user
+     * @throws Throwable
+     */
+    public static function sendDeactivationEmail(User $user)
+    {
+
+        $title = "Account subject to deactivation";
+        $message = "Your account is subject to deactivation due to your inactivity in TrenchDevs. Please login within the next 3 days to avoid this.";
+
+        $viewData = [
+            'name' => $user->name(),
+            'email_body' => $message,
+        ];
+
+        EmailQueue::queue(
+            trim($user->email),
+            $title,
+            $viewData,
+            'emails.generic'
+        );
+
+    }
+
+    /**
+     * Deactivates inactive users even after notice period
+     *
+     * @param int $noticeDays
+     * @param string $userRole
+     */
+    public static function deactivateUsers(int $noticeDays = 3, string $userRole = self::ROLE_CONTRIBUTOR)
+    {
+
+        $latestLogins = DB::table('user_logins')
+            ->selectRaw('user_id, MAX(created_at) AS last_login')
+            ->groupBy('user_id');
+
+        // Deactivate flagged users who DID NOT sign in within 3 days from deactivation notice
+        DB::table('users')
+            ->joinSub($latestLogins, 'latest_logins', function ($join) {
+                $join->on('users.id', '=', 'latest_logins.user_id');
+            })
+            ->where('users.is_active', 1)
+            ->where('users.is_flagged_for_deactivation', 1)
+            ->where('users.role', $userRole)
+            ->whereRaw('TIMESTAMPDIFF(DAY, users.deactivation_notice_sent_at, NOW()) > ?', $noticeDays)
+            ->whereRaw('TIMESTAMPDIFF(DAY, latest_logins.last_login, NOW()) > ?', $noticeDays)
```

Comment:
can use php here instead of mysql (easier to read & faster), eg:
```
->where(users.deactivation_notice_sent_at, '>', '<<generate 3 days ago [strtotime]>>')
```

Reply:
For discussion